### PR TITLE
Proc macros

### PR DESCRIPTION
**Pull request by saolsen**
Works. I need to clean up some of the attribute finding code and decide if bad attributes are errors or just ignored. There's some extra match branches in there just because I sort of expect those paths to be useful when we add more attributes and I don't want to forget about them.

Right now it lets you specify the class name and which fields you want getters for.
It derives a few functions on an interface, one that returns a full class definition and one that returns a partially filled out builder that you can then add more methods to.

I think it's almost good enough for v1. I have to make constructors be their own builder method since we can't get them up front from an attribute (yet). Right now it defaults to `new` but we still want people to be able to specify a custom one.

Over all this went pretty well.

### CODE REVIEW

File: languages/rust/oso/tests/test_polar.rs [hunk 1]
```diff
@@ -244,3 +245,43 @@ fn test_external() {
...
+        .name("Baz")
+        .add_method("world", |receiver: &Foo| format!("{} world", receiver.a))
+        .build();
+    test.oso.register_class(class).unwrap();
+
+    test.qvar_one(
+        r#"new Baz("hello").world() = x"#,
+        "x",
+        "hello world".to_string(),
+    );
```
Comment (saolsen) [inline languages/rust/oso/tests/test_polar.rs:286]: Here's what the usage looks like.

File: languages/rust/oso/src/host/class.rs [hunk 2]
```diff
@@ -160,13 +178,17 @@ impl<T> Class<T> {
...
+        if let Some(constructor) = &self.constructor {
+            let instance = constructor.invoke(fields, host)?;
+            Ok(Instance {
+                name: self.name.clone(),
+                instance,
+                attributes: Arc::new(self.attributes.clone()),
+                methods: Arc::new(self.instance_methods.clone()),
+            })
+        } else {
+            panic!("Invalid class, should already have a constructor")
```
Comment (samscott89) [inline languages/rust/oso/src/host/class.rs:190]: Classes don't _have_ to be constructable, right? No constructor seems okay, should be a regular error.

Comment (saolsen) [reply languages/rust/oso/src/host/class.rs:190]: Oh true, cool.

File: languages/rust/oso-derive/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+                    get_nested_attr(nested, oso_attrs);
+                }
+            }
+            _ => (),
+        }
+    }
+}
+
+#[proc_macro_derive(PolarClass, attributes(polar))]
+pub fn derive_testing_fn(ts: TokenStream) -> TokenStream {
```
Comment (dhatch) [inline languages/rust/oso-derive/src/lib.rs:79]: Why this function name?

**Review by dhatch — APPROVED** [review]
lgtm, todo clean up as follow up

File: languages/rust/oso/tests/test_polar.rs [hunk 4]
```diff
@@ -244,3 +245,46 @@ fn test_external() {
...
+#[test]
+fn test_macros() {
+    // stub
+
+    let _ = tracing_subscriber::fmt::try_init();
+
+    #[derive(PolarClass)]
+    #[polar(class_name = "Bar")]
+    struct Foo {
+        #[polar(attribute)]
```
[the remark this answers is not in the document]
Comment (saolsen) [reply languages/rust/oso/tests/test_polar.rs:315]: Added

File: languages/rust/oso-derive/src/lib.rs (hunk 3 above)
Comment (saolsen) [reply languages/rust/oso-derive/src/lib.rs]: old name, fixed

**Pull request merged**